src/main/java/org/openstreetmap/atlas/checks/validation/intersections/EdgeCrossingEdgeCheck.java

```diff
@@ -191,4 +117,93 @@ public boolean validCheckForObject(final AtlasObject object)
     {
         return FALLBACK_INSTRUCTIONS;
     }
+
+    /**
+     * Validates given {@link AtlasObject} (assumed to be an {@link Edge}) whether it is a valid
+     * crossing edge or not
+     *
+     * @param object
+     *            {@link AtlasObject} to test
+     * @return {@code true} if given {@link AtlasObject} object is a valid crossing edge
+     */
+    private boolean isValidCrossingEdge(final AtlasObject object)
+    {
+        if (Edge.isMasterEdgeIdentifier(object.getIdentifier())
+                && !TagPredicates.IS_AREA.test(object) && !this.isFlagged(object.getIdentifier()))
+        {
+            final Optional<HighwayTag> highway = HighwayTag.highwayTag(object);
+            if (highway.isPresent())
+            {
+                final HighwayTag highwayTag = highway.get();
+                return HighwayTag.isCarNavigableHighway(highwayTag)
+                        && !HighwayTag.CROSSING.equals(highwayTag)
+                        && highwayTag.isMoreImportantThanOrEqualTo(this.minimumHighwayType);
+            }
+        }
+        return false;
+    }
+
+    /**
+     * A BFS walker that collects all connected invalid crossings. The use of the walker prevents
+     * edges from being flagged mor than once.
+     *
+     * @param startingEdge
+     *            {@link Edge} to start the walker from.
+     * @return a {@lionk Set} of invalid crossing {@link Edge}s
+     */
+    private Set<Edge> getInvalidCrossingEdges(final Edge startingEdge)
+    {
+        final Set<Edge> invalidEdges = new HashSet<>();
+        final Queue<Edge> toCheck = new ArrayDeque<>();
+        invalidEdges.add(startingEdge);
+        toCheck.add(startingEdge);
+
+        while (!toCheck.isEmpty())
+        {
+            final Edge edge = toCheck.poll();
+            // Prepare the edge being tested for checks
+            final PolyLine edgeAsPolyLine = edge.asPolyLine();
+            final Rectangle edgeBounds = edge.bounds();
+            // If layer tag is present use its value, else use the OSM default
+            final Optional<Long> edgeLayer = Validators.hasValuesFor(edge, LayerTag.class)
```

Comment:
The reason I did it this way is so that if there is no layer tag it has the default, but if there is it still returns the optional (which is empty if the layer tag is not a number in the range -5 to 5).